File: app/src/main/java/com/mdgd/pokemon/bg/LoadPokemonsModel.java
package com.mdgd.pokemon.bg;

import com.mdgd.pokemon.models.cache.Cache;
import com.mdgd.pokemon.models.infra.Result;
import com.mdgd.pokemon.models.repo.PokemonsRepo;

import java.util.concurrent.TimeUnit;

import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class LoadPokemonsModel implements LoadPokemonsContract.ServiceModel {

    private CompositeDisposable disposables = new CompositeDisposable();
    private final PokemonsRepo repo;
    private final Cache cache;

    public LoadPokemonsModel(PokemonsRepo repo, Cache cache) {
        this.repo = repo;
        this.cache = cache;
    }

    @Override
    public void load() {
        disposables.addAll(
                // complete task, dispose all subscriptions
                cache.getProgressObservable()
                        .filter(event -> event.getValue() == -1)
                        .delay(50, TimeUnit.MILLISECONDS, Schedulers.trampoline())
                        .subscribe(event -> disposables.clear()),

                repo.loadPokemons()
                        .doFinally(() -> cache.putLoadingProgress(new Result<>(PokemonsRepo.LOADING_COMPLETE)))
                        .subscribe(cache::putLoadingProgress, error -> cache.putLoadingProgress(new Result<>(error)))
        );
    }
}

File: app/src/main/java/com/mdgd/pokemon/bg/LoadPokemonsModel.kt
package com.mdgd.pokemon.bg

import com.mdgd.pokemon.models.cache.Cache
import com.mdgd.pokemon.models.infra.Result
import com.mdgd.pokemon.models.repo.PokemonsRepo
import kotlinx.coroutines.*

class LoadPokemonsModel(private val repo: PokemonsRepo, private val cache: Cache) : ServiceModel {
    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val exceptionHandler = CoroutineExceptionHandler { _, e ->
        cache.putLoadingProgress(Result(e))
        coroutineScope.cancel()
    }

    override fun load() {
        coroutineScope.launch(exceptionHandler) {
            val initialAmount = PokemonsRepo.PAGE_SIZE.toLong() * 2
            repo.loadInitialPages(initialAmount)
            cache.putLoadingProgress(Result(initialAmount))

            val newAmount = repo.loadPokemons(initialAmount)
            cache.putLoadingProgress(Result(newAmount))
            coroutineScope.cancel()
        }
    }
}
